services/notification_queue/notification_queue.py

```diff
@@ -0,0 +1,1115 @@
+client = None
+if account_sid and auth_token and from_number:
+    client = TwilioClient(account_sid, auth_token)
+
+email = False
+if cfg["email_service"] == "sendgrid" or cfg["email_service"] == "smtp":
+    email = True
+
+
+op_options = [
+    "lt",
+    "le",
+    "eq",
+    "ne",
+    "ge",
+    "gt",
+]
+
+
+def notification_resource_type(target):
+    if not target.notification_type:
+        return None
+    if "favorite_sources" not in target.notification_type:
+        return target.notification_type
+    elif "favorite_sources" in target.notification_type:
+        return "favorite_sources"
+
+
+def user_preferences(target, notification_setting, resource_type):
+
+    if not isinstance(notification_setting, str):
+        return
+    if not isinstance(resource_type, str):
+        return
+    if not target.user:
+        return
+
+    if notification_setting == "email":
+        if not email:
+            return
+        if not target.user.contact_email:
+            return
+        # this ensures that an email is sent regardless of the user's preferences
+        # this is useful for group_admission_requests, where we want the admins to always be notified by email
+        if resource_type in ['group_admission_request']:
+            return True
+
+    if not target.user.preferences:
+        return
+
+    if notification_setting in ["sms", "phone"]:
+        if client is None:
+            return
+        if not target.user.contact_phone:
+            return
+
+    if notification_setting == "slack":
+        if not target.user.preferences.get('slack_integration'):
+            return
+        if not target.user.preferences['slack_integration'].get("active"):
+            return
+        if (
+            not target.user.preferences['slack_integration']
+            .get("url", "")
+            .startswith(cfg["slack.expected_url_preamble"])
+        ):
+            return
+
+    prefs = target.user.preferences.get('notifications')
+    if not prefs:
+        return
+    else:
+        if resource_type in [
+            'sources',
+            'favorite_sources',
+            'gcn_events',
+            'facility_transactions',
+            'mention',
+            'analysis_services',
+            'observation_plans',
+        ]:
+            if not prefs.get(resource_type, False):
+                return
+            if not prefs[resource_type].get(notification_setting, False):
+                return
+            if not prefs[resource_type][notification_setting].get("active", False):
+                return
+
+        return prefs
+
+
+def send_slack_notification(session, target):
+    resource_type = notification_resource_type(target)
+    notifications_prefs = user_preferences(target, "slack", resource_type)
+    if not notifications_prefs:
+        return
+    integration_url = target.user.preferences['slack_integration'].get('url')
+
+    slack_microservice_url = f'http://127.0.0.1:{cfg["slack.microservice_port"]}'
```

Comment:
here we might want to use the app URL and port instead. Not sure how this behaves in a docker-compose network (idk if everything is on localhost or not).

Reply:
This is also how we do the other microservices.